era_5g_interface/task_handler_internal_q.py

```diff
@@ -45,13 +67,40 @@ def store_image(self, metadata: dict, image: np.ndarray) -> None:
                 The format is NetApp-specific.
             image (_type_): The image to be processed.
         """
+
+        # make sure multiple calls of store_image do not interfere with each other
+        self._q_put_lock.acquire()
+
         self.frame_id += 1
         # logger.info(f"TaskHandlerInternalQ received frame id: {self.frame_id} with timestamp: {metadata['timestamp']}")
         try:
             self._q.put((metadata, image), block=False)
         except Full:
-            pass
-            # TODO: raise an exception
+            if self.if_queue_full == "discard_new":
+                pass  # do nothing (discard the new data)
+            elif self.if_queue_full == "discard_oldest":
+                # remove the oldest data from queue to make up space for the new one,
+                # but make sure not to discard control commands
+                self._q.mutex.acquire()  # acquire the queue lock to prevent it from mutating during the removal
```

Comment:
Isn't one lock (`self._q_put_lock`) enough? Or couldn't we use just the "queue lock"?

Reply:
Sure, we can meet, but if it is essential, it should be fully explained in the source code so we will know the reason even next week or next month :-)